builtin/logical/pki/cert_util.go

```diff
@@ -162,6 +220,237 @@ func validateCommonNames(req *logical.Request, commonNames []string, role *roleE
+	signingBundle *certutil.ParsedCertBundle,
+	pkiAddress string,
+	req *logical.Request,
+	data *framework.FieldData) (*certutil.ParsedCertBundle, error) {
+
+	csrString := req.Data["csr"].(string)
+	if csrString == "" {
+		return nil, certutil.UserError{Err: fmt.Sprintf(
+			"\"csr\" is empty")}
+	}
+
+	pemBytes := []byte(csrString)
+	pemBlock, pemBytes := pem.Decode(pemBytes)
+	if pemBlock == nil {
+		return nil, certutil.UserError{Err: "csr contains no data"}
+	}
+	csr, err := x509.ParseCertificateRequest(pemBlock.Bytes)
+	if err != nil {
+		return nil, certutil.UserError{Err: "certificate request could not be parsed"}
+	}
+
+	creationBundle, err := generateCreationBundle(b, role, signingBundle, pkiAddress, req, data)
+	if err != nil {
+		return nil, err
+	}
+
+	parsedBundle, err := signCertificate(creationBundle, csr)
+	if err != nil {
+		return nil, err
+	}
+
+	return parsedBundle, nil
+}
+
+func generateCreationBundle(b *backend,
+	role *roleEntry,
+	signingBundle *certutil.ParsedCertBundle,
+	pkiAddress string,
+	req *logical.Request,
+	data *framework.FieldData) (*certCreationBundle, error) {
+	var err error
+
+	// Get the common name(s)
+	cn := data.Get("common_name").(string)
+	if len(cn) == 0 {
+		return nil, certutil.UserError{Err: "The common_name field is required"}
+	}
+
+	dnsNames := []string{}
+	emailAddresses := []string{}
+	if strings.Contains(cn, "@") {
+		emailAddresses = append(emailAddresses, cn)
+	} else {
+		dnsNames = append(dnsNames, cn)
+	}
+	cnAltInt, ok := data.GetOk("alt_names")
+	if ok {
+		cnAlt := cnAltInt.(string)
+		if len(cnAlt) != 0 {
+			for _, v := range strings.Split(cnAlt, ",") {
+				if strings.Contains(v, "@") {
+					emailAddresses = append(emailAddresses, cn)
+				} else {
+					dnsNames = append(dnsNames, v)
+				}
+			}
+		}
+	}
+
+	// Get any IP SANs
+	ipAddresses := []net.IP{}
+	ipAltInt, ok := data.GetOk("ip_sans")
+	if ok {
+		ipAlt := ipAltInt.(string)
+		if len(ipAlt) != 0 {
+			if !role.AllowIPSANs {
+				return nil, certutil.UserError{Err: fmt.Sprintf(
+					"IP Subject Alternative Names are not allowed in this role, but was provided %s", ipAlt)}
+			}
+			for _, v := range strings.Split(ipAlt, ",") {
+				parsedIP := net.ParseIP(v)
+				if parsedIP == nil {
+					return nil, certutil.UserError{Err: fmt.Sprintf(
+						"the value '%s' is not a valid IP address", v)}
+				}
+				ipAddresses = append(ipAddresses, parsedIP)
+			}
+		}
+	}
+
+	var ttlField string
+	ttlFieldInt, ok := data.GetOk("ttl")
+	if !ok {
+		ttlField = role.TTL
+	} else {
+		ttlField = ttlFieldInt.(string)
+	}
+
+	var ttl time.Duration
+	if len(ttlField) == 0 {
+		ttl = b.System().DefaultLeaseTTL()
+	} else {
+		ttl, err = time.ParseDuration(ttlField)
+		if err != nil {
+			return nil, certutil.UserError{Err: fmt.Sprintf(
+				"invalid requested ttl: %s", err)}
+		}
+	}
+
+	var maxTTL time.Duration
+	if len(role.MaxTTL) == 0 {
+		maxTTL = b.System().MaxLeaseTTL()
+	} else {
+		maxTTL, err = time.ParseDuration(role.MaxTTL)
```

Comment:
That is true, but not all paths to calling this function use a roleEntry saved via the normal path; in particular the paths that generate CA certs don't use saved roles. I could move the logic out there as well, but figured it's better to keep it centralized rather than forget to duplicate it.

The check in `path_roles.go` is now redundant, but has the useful property of erroring out when they're trying to save a role, rather than saving the role and erroring out when trying to generate certs.

That said, this did serve as a good reminder that I need to only actually save a value for the role if they set it...